pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -707,9 +731,11 @@ public void setReadOnly(boolean readOnly) throws SQLException {
     }
 
     if (readOnly != this.readOnly) {
-      String readOnlySql
-             = "SET SESSION CHARACTERISTICS AS TRANSACTION " + (readOnly ? "READ ONLY" : "READ WRITE");
-      execSQLUpdate(readOnlySql); // nb: no BEGIN triggered.
+      //if autocommit is true, then we change things at the session level
+      //if autocommit is false, read only is managed with the transaction
+      if (autoCommit) {
+        execSQLUpdate(readOnly ? setSessionReadOnly : setSessionNotReadOnly);
```

Comment:
Just to make sure I understand. You are proposing a new connection property which would, by default, make setting readOnly only used when autoCommit is false?

This addresses issue 1228 by not doing any work when setReadOnly is called.

When autocommit is false transaction will be read only, which gives hint to pools.

Reply:
Well, it might make sense to have it a tri-state option (default should be like item 2):
1) always ignore readonly
2) pass readonly for begin only (autocommit==false)
3) pass readonly always (like 2, but with set session in autocommit mode)

The idea is to avoid overheads, and still pass hints downstream.

Reply:
Is your option 3 the behavior in 42.2.0 or the behavior currently in this PR?

Is there value in 4 options?
IGNORE
TRANSACTION
MIXED (currently this PR)
SESSION (current behavior in 42.2.0)

Reply:
this PR.

I think we can abandon 42.2.0 behavior (==always fuse `begin readonly`)

Reply:
And have default behavior be TRANSACTION?

Reply:
yes